Sources/ArgumentParser/Usage/UsageGenerator.swift

```diff
@@ -339,16 +339,18 @@ extension ErrorMessageGenerator {
     }
   }
   
-  func unableToParseValueMessage(origin: InputOrigin, name: Name?, value: String, key: InputKey) -> String {
+    func unableToParseValueMessage(origin: InputOrigin, name: Name?, value: String, key: InputKey, message: String?) -> String {
     let valueName = arguments(for: key).first?.valueName
-    switch (name, valueName) {
-    case let (n?, v?):
+    switch (name, valueName, message) {
+    case let (_, _, m?):
+      return "\(m)"
+    case let (n?, v?, _):
       return "The value '\(value)' is invalid for '\(n.synopsisString) <\(v)>'"
-    case let (_, v?):
+    case let (_, v?, _):
       return "The value '\(value)' is invalid for '<\(v)>'"
-    case let (n?, _):
+    case let (n?, _, _):
       return "The value '\(value)' is invalid for '\(n.synopsisString)'"
-    case (nil, nil):
+    case (nil, nil, _):
       return "The value '\(value)' is invalid."
     }
```

Comment:
I suppose it's a matter of preference, but I feel like we should just keep the context, and append the custom message if it is not `nil`. This would also remove the need for `CustomParserErrorConvertible` and instead just use `localizedDescription`.

Reply:
My preference is not to append custom messages to any default error or context. It's not really a custom message at that point, if I'm customising the error I want full control.

But I agree the context is useful. And using something like `localizedDescription` would make it cleaner. 

Maybe we should consider wrapping the original error (thrown from the transform closure) in the `unableToParseValue` instead of a custom error message. 

But then we'd need some error formatting option, possibly in `CommandConfiguration`?

Reply:
I agree with @jayrhynas here — the printed error for this should combine the existing error output with a string from the thrown error. The output for a transformation throws on non-alphanumeric input could look like this:

```
$ command --option bad-input-#!@#
Error: The value 'bad-input' is invalid for '--option': Contains invalid characters '#!@#'
Usage: command [--option <option>] ...
``` 

where `Contains invalid characters '#!@#'` is the string representation of the thrown error.

> Maybe we should consider wrapping the original error (thrown from the transform closure) in the unableToParseValue instead of a custom error message.

This sounds good! We don't want to add a protocol just for this functionality, so doing a best effort conversion of the error to a string would be great. In general we try to do that by trying to match on `LocalizedError where error.errorDescription != nil` and then just falling back to `String(describing: error)`. That makes it simple for people to have nice error messages by just conforming their error type to `CustomStringConvertible`.

Reply:
I have updated `unableToParseValue` to take an optional `Error` associated value. 

If this error is not nil `unableToParseValueMessage(…)` makes best-efforts to create a custom error message.

> The output for a transformation throws on non-alphanumeric input could look like this:
> 
> ```
> $ command --option bad-input-#!@#
> Error: The value 'bad-input' is invalid for '--option': Contains invalid characters '#!@#'
> Usage: command [--option <option>] ...
> ```
> 
> where Contains invalid characters '#!@#' is the string representation of the thrown error.

This is exactly how the errors are now formatted. "Contains invalid characters '#!@#'" must be returned by either `errorDescription` form a `LocalizedError` or `description` from `CustomStringConvertible`. If the error is `nil` the errors are formatted as before.

`ValidationError` works as expected, printing the `message` string.

Simple `enum`s, such as:

```swift
enum FooError: Error {
    case outOfBounds
    case unknown 
}
```

Simply print out the name of the `case`.